### DISCUSSION TITLE
Update the wiki wm8960 page

### DISCUSSION THREAD

**Discussion by mittler-works** [MiczFlor/RPi-Jukebox-RFID#2046] [Q&A]
Hi, I have a suggestion to improve the wiki.

Atm I build a Phoniebox using the wm8960 soundcard. Neither the WaveShare driver nor the proposed seeed-voicecard driver will work on newer Raspberry Pi OS Releases (Bullseye+).

After pretty much pain I found out, there is support for this soundcard in the raspberry firmware! Have a look [here](https://github.com/raspberrypi/firmware/tree/master/boot/overlays)!

I somehow get it to work by adding
```
dtparam=i2c_arm=on
dtparam=i2s=on
dtparam=spi=on
dtoverlay=i2s-mmap
dtoverlay=wm8960-soundcard
```
to the `/boot/config.txt`-file.

Afterwards I had to copy the `asound.conf` from Waveshare's Repo [here](https://github.com/waveshare/WM8960-Audio-HAT/blob/master/asound.conf) to `/etc/asound.conf` and the `wm8960_asound.state` to `/var/lib/alsa/asound.state`.

Then I installed (via apt) `samplerate-programs` `i2c-tools` `libasound2-plugins`

After a reboot I could use the card. I think that's the most clean and easy way to use this card, BUT: I am not sure which of the above steps and options are really necessary and which are not.
I would like to find out, or maybe one of you knows, but after that I would like to update the wiki and establish this way as the "standard way" for the wm8960 sound card.

What do you think?

**Comment by mittler-works** [comment]
I'm building my second Phoniebox now, again with the wm8960 soundcard.

It seems, for bullseye, it is enough to:

1. copy the `asound.conf` as mentioned above
2. to install the `libasound2-plugins` Package
3. and to add these to your config.txt:
```
dtparam=i2c=on
dtparam=i2s=on
dtoverlay=wm8960-soundcard
```

Nothing more.

However this does not work on bookworm anymore! I'm investigating this problem and if I find a good way that works for bookworm too, I'll update the wiki accordingly.

**Reply by s-martin** [reply]
Hi,
there are some issues on Bookworm (see also #2295).

Currently the recommendation is to use only **Bullseye 32bit Lite**.

If you find solutions to fix the issues under Bookworm these are more than welcome!

**Reply by mittler-works** [reply]
Thank you very much for your answer! I've already seen this ticket and still wanted to use bookworm, because.. well.. there must be some crazy people who like to go through pain and suffering to find solutions. ;)

On Topic. I've found a solution but I am not sure about the cause. The required drivers, codecs and device tree overlays are still there in bookworm. It just seems to be broken somewhere. I'm investigating on that and I'll open an issue in [raspberry/linux](https://github.com/raspberrypi/linux) as soon as I found some more details.

In the meantime, I noticed that **Waveshare has updated [their repo](https://github.com/waveshare/WM8960-Audio-HAT)!** To be honest, I've had abandoned all hope, that this would ever happen. That's why I didn't even look at Waveshares repo in the first place. But they did and it works!

So you **could** just simply do it like they describe on their readme and it would work (if you are on a fresh install, you may install git first):
```
git clone https://github.com/waveshare/WM8960-Audio-HAT
cd WM8960-Audio-HAT
sudo ./install.sh 
sudo reboot
```

That's great for now, but I do not really like this solution. In my experience it is almost always better not to self build kernel modules because sooner and later things will break. Even though dkms takes care of recompiling modules on kernel upgrades, it is always a little bit hacky.

Furthermore the Waveshare install script installs some unnecessary stuff like their wm8960-soundcard.service, clutters your config files with unnecessary and deprecated options like i2s-mmap and overwrites the firmware wm8960-soundcard.dtbo with an outdated one. The only thing that you really need from waveshare is the driver and the asound.conf.

So if you want to keep your installation clean, just add an `exit` statement right after the function call `install_module "./" "wm8960-soundcard"` [here](https://github.com/waveshare/WM8960-Audio-HAT/blob/259d282dc4856318c6961e7d6a94b761308653d3/install.sh#L51) in the install script before you run it.

Then all you have to do is copy the `asound.conf` to `/etc/asound.conf` and add `dtoverlay=wm8960-soundcard` to your `/boot/firmware/config.txt`. You don't need no service. The dependent device tree overlays (i2c, i2s) will be loaded implicitly. Also the required kernel modules load implicitly, no need to edit /etc/modules.

---

For the record. Journalctl shows me the error `No MCLK configured` with the official raspberry drivers. So there might be some issue when configuring the clocks.

The official codec and driver file pendants to the waveshare codec and driver files are:

| Waveshare | Raspberry |
| --- | --- |
| [wm8960.c](https://github.com/waveshare/WM8960-Audio-HAT/blob/master/wm8960.c) | [wm8960.c](https://github.com/raspberrypi/linux/blob/rpi-6.6.y/sound/soc/codecs/wm8960.c) |
| [wm8960.h](https://github.com/waveshare/WM8960-Audio-HAT/blob/master/wm8960.h) | [wm8960.h](https://github.com/raspberrypi/linux/blob/rpi-6.6.y/sound/soc/codecs/wm8960.h) |
| [wm8960-soundcard.c](https://github.com/waveshare/WM8960-Audio-HAT/blob/master/wm8960-soundcard.c) | [simple-card.c](https://github.com/raspberrypi/linux/blob/rpi-6.6.y/sound/soc/generic/simple-card.c) |

As you can see when you diff those files, Waveshare drops some of the parameterization needed for the driver to work with different vendors that use the wm8960 codec. They skip parts of initialization and just hardcode some of the clock configuration. In the header file they simply swap the makro for the default clock. So it is no option to merge these changes into the raspberry linux.

---

For now, I would update the wiki with the information that Waveshare had updated their drivers to work with bookworm. As soon as the official raspberry driver is fixed, I would go for the no-third-party-option.

**Reply by s-martin** [reply]
In #2313 there is some info about making Phoniebox work under Bookworm (even 64 bit)
